### PR TITLE
Output filtering

### PR DESCRIPTION
**Pull request by ashh87**
This set of patches extends support for helpers by adding an option to specify whether they are added to the raw OUTPUT chain or the raw PREROUTING chain. I'm not sure if there is a situation where both might be required for a helper, but I can update if so. If unspecified, the default is PREROUTING, so existing helper xml files are currently unchanged, and custom helpers outside of this repo won't break.

This might help with resolving #185 .The SSDP conntrack helper needs to go on raw OUTPUT, so this is needed for #25.

### CODE REVIEW

**Comment by t-woerner** [comment]
Please append new fields always to the end of a D-Bus signature. Inserting them in the middle is not recommended as it will break existing code.
I think that a field as `direction` (`input`|`output`) might be better as naming the chain.

File: src/firewall/core/io/helper.py [hunk 1]
```diff
@@ -42,17 +42,18 @@ class Helper(IO_Object):
         ( "description", "" ),                # s
         ( "family", "", ),                    # s
         ( "module", "", ),                    # s
+        ( "chain", "", ),                     # s
         ( "ports", [ ( "", "" ), ], ),        # a(ss)
         )
-    DBUS_SIGNATURE = '(sssssa(ss))'
+    DBUS_SIGNATURE = '(ssssssa(ss))'
```
Comment (t-woerner) [inline src/firewall/core/io/helper.py]: Please add the new element to the end of the signature.

**Comment by ashh87** [comment]
I've updated so that UI is setting the 'direction' property to either 'input' or 'output'.

**Comment by ashh87** [comment]
Would this be more useful if I merged it with master?

**Comment by erig0** [comment]
It would be helpful if you could squash some commits together. There are quite a few instances where something changes, then changes again in later patches. This makes it hard to review. For example, the dbus signature changes 3 times in separate commits.

**Comment by ashh87** [comment]
No problem, I'll get that done soon.

**Comment by ashh87** [comment]
Hopefully that should be a bit simpler, although github pull request shows the commits in the wrong order. I've not squashed the renaming of 'chain' to 'direction' as suggested, because it seems like a separate issue to me, and because it required a new dictionary to be added to convert one string to another. Please let me know if it needs further improvements!

**Comment by erig0** [comment]
@ashh87, sorry for the long delay. Thanks again for your effort on these patches.

I have been giving this quite a bit of thought. There are also many requests to support OUTPUT filtering in other areas of firewalld. As such, I think this could be better implemented by adding a generic "--direction={input, output}" flag. This could support all aspects of a service; rules, helper, etc. It also be used for --add-protocol, --add-port, etc. The direction flag could also be extended to support "forward" in the future.

I've begun working on this, but it is not a small task and likely means introducing new DBUS methods. I will update this issue when I have something more to report.

**Comment by erig0** [comment]
Closing this PR. We need to implement general OUTPUT/FORWARD filtering support not helper specific extensions.

**Comment by henryptung** [comment]
@erig0 Is there a general issue tracking this?

**Comment by erig0** [comment]
> Is there a general issue tracking this? 

Yes, #32.

> FORWARD filtering is its own concept, which involves a lot of new design

Indeed. Awhile ago I put together proposed changes that cover all the points you mention. I'll have to clean that up and push it to the git repo.

> OUTPUT filtering could be useful, but I think one of its features will be exactly the "helper-specific extension" you mention above.

There are other use cases outside of helpers; e.g. notify of new traffic from applications.

**Comment by mwilck** [comment]
I just became aware of the issue here. Let me summarize shortly, correct me if I'm wrong.

The `samba-client` service of firewalld is non-functional. It opens port 137/udp and 138/udp, which is not necessary for SMB client functionality, but blocks responses from UDP port 137 (see #266, #267). There are two possible workarounds:

 1. open source port 137
 2. enable "automatic helpers" with `samba-client`, or enable the `net.netfilter.nf_conntrack_helper` sysctl system-wide.

*Both workarounds are discouraged for security reasons*. (2) is probably less critical, so it would be recommended for end users who need SMB client functionality (?)

The reason these workarounds are necessary is that firewalld implements the conntrack filtering logic wrongly, at least for `netbios-ns`. It inserts a `-j CT --helper netbios-ns` rule in the `PREROUTING` chain, but that's wrong - the rule belongs in the `OUTPUT` chain. The rule in `PREROUTING` is *not* necessary to make NetBIOS name lookups work.

This problem has been known for at least 3 years, and patches have been proposed e.g. in this PR, but have been rejected in favor of a more generalized approach which is tracked in #32. The latter issue is tracked as an RFE, as it's mainly seen as an additional feature. Very obviously it has been tracked with priority, which is appropriate for an RFE, but not so much for a bug. 

Unless I'm missing something essential, the `samba-client` service in firewalld is broken, and I would call that a bug. I believe that some short-term solution should be targeted for this problem, as the generic "output filtering" functionatlity is apparently still quite far away. There are likely other conntrack helpers that are affected by this the same way as `netbios-ns`.

**Pull request closed without merging**